### PR TITLE
fix: dataEndIndex can be -1

### PR DESCRIPTION
**Pull request by HHongSeungWoo**
## Description

As dataEndIndex is -1 in an empty data array, it causes issues with #3987.
It doesn't seem to be intentional

## Related Issue

#3987

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/chart/generateCategoricalChart.tsx [hunk 1]
```diff
@@ -649,7 +649,9 @@ const createDefaultState = (props: CategoricalChartProps): CategoricalChartState
   const brushItem = findChildByType(children, Brush);
   const startIndex = (brushItem && brushItem.props && brushItem.props.startIndex) || 0;
   const endIndex =
-    brushItem?.props?.endIndex !== undefined ? brushItem?.props?.endIndex : (props.data && props.data.length - 1) || 0;
+    brushItem?.props?.endIndex !== undefined
+      ? brushItem?.props?.endIndex
+      : (props.data && props.data.length !== 0 && props.data.length - 1) || 0;
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:654]: This approach makes this section of code look really confusing imo. If we moved the check and the resulting value outside of the ternary this becomes so much easier to read

**Review by ckifer — CHANGES_REQUESTED** [review]

**Review by nikolasrieble — COMMENTED** [review]
Could we add a test to this change? I see how a unit test can be difficult, but a RTL or storybook test should be possible.

**Comment by HHongSeungWoo** [comment]
I will also add a test

File: test/chart/generateCategoricalChart.spec.tsx [hunk 2]
```diff
@@ -230,4 +231,27 @@ describe('generateCategoricalChart', () => {
       );
     });
   });
+
+  describe('createDefaultState', () => {
+    it('Should have the correct dataIndex', () => {
```
Comment (nikolasrieble) [inline test/chart/generateCategoricalChart.spec.tsx:236]: Why did you combine these four separate test cases into a single test? Is there any interaction that I am missing? If not, please split it up into 4 tests.

Comment (HHongSeungWoo) [reply test/chart/generateCategoricalChart.spec.tsx:236]: it was just ctrl+c,v :sweat_smile:

**Review by nikolasrieble — APPROVED** [review]
Thank you for adding the tests! ❤️ 🚀

**Review by ckifer — APPROVED** [review]

**Pull request merged**